### PR TITLE
[HiCache] Optimize LogicalHostPool free-list release

### PR DESCRIPTION
**Pull request by wangwenmingaa**
<html><head></head><body><h2><span>Motivation</span></h2><p><span>This PR is a follow-up to </span><a href="https://github.com/sgl-project/sglang/pull/30658"><span>#30658</span></a><span>.</span></p><p><span>PR #30658 defers host-pool free-list concatenation for </span><code><span>HostKVCache</span></code><span>-based pools. However, </span><code><span>LogicalHostPool</span></code><span> is a standalone allocator and does not inherit from </span><code><span>HostKVCache</span></code><span>, so it is not covered by that optimization.</span></p><p><code><span>LogicalHostPool.free()</span></code><span> still eagerly rebuilds the complete free-slot tensor on every release:</span></p><pre><code><span>self.free_slots = torch.cat([self.free_slots, released_indices])</span></code></pre><p><span>When HiCache uses a large logical host pool, </span><code><span>free_slots</span></code><span> can contain hundreds of thousands or millions of entries. Releasing a small number of slots therefore copies the entire existing free list. Repeated releases during cache eviction can introduce significant CPU stalls during HiCache cold fill and warm-up.</span></p><p><span>In an internal DSV4 HiCache workload, the allocator eviction-and-retry path showed representative average latencies of approximately </span><code><span>392.6–586.7 ms</span></code><span>, with a maximum close to </span><code><span>592 ms</span></code><span>. Profiling identified repeated large free-list concatenations as the dominant CPU-side overhead.</span></p><p><span>The issue was especially visible with a large logical cache capacity, but the underlying allocator problem is independent of any particular cache-sharding configuration.</span></p><h2><span>Modifications</span></h2><p><span>Apply the lazy free-slot release strategy from #30658 to </span><code><span>LogicalHostPool</span></code><span>:</span></p><ul><li><p><span>Keep released slot tensors in </span><code><span>release_slots</span></code><span> instead of concatenating them immediately.</span></p></li><li><p><span>Track pending released slots with </span><code><span>num_release_slots</span></code><span>.</span></p></li><li><p><span>Include pending released slots in </span><code><span>available_size()</span></code><span>.</span></p></li><li><p><span>Continue allocating from the primary </span><code><span>free_slots</span></code><span> tensor while it has sufficient capacity.</span></p></li><li><p><span>Merge pending released chunks only when an allocation cannot be satisfied by the primary free list.</span></p></li><li><p><span>Avoid recording empty releases.</span></p></li><li><p><span>Reset all pending-release state in </span><code><span>clear()</span></code><span>.</span></p></li><li><p><span>Preserve the existing page-alignment validation and slot reuse order.</span></p></li></ul><p><span>The patch does not add runtime logging, configuration flags, or model-specific behavior.</span></p><p><span>Unit tests cover:</span></p><ul><li><p><span>deferred release without rebuilding the primary free list;</span></p></li><li><p><span>allocation from the primary free list before merging pending slots;</span></p></li><li><p><span>reuse of released slots after a deferred merge;</span></p></li><li><p><span>merging multiple released chunks;</span></p></li><li><p><span>empty releases;</span></p></li><li><p><span>state reset through </span><code><span>clear()</span></code><span>;</span></p></li><li><p><span>preservation of </span><code><span>LogicalHostPool</span></code><span> page-alignment checks.</span></p></li></ul><h2><span>Accuracy Tests</span></h2><p><span>This change only modifies CPU-side free-slot allocator metadata.</span></p><p><span>It does not modify:</span></p><ul><li><p><span>model weights or model forward execution;</span></p></li><li><p><span>KV cache tensor contents;</span></p></li><li><p><span>cache keys or page-table mappings;</span></p></li><li><p><span>host-to-device or device-to-host data transfer;</span></p></li><li><p><span>attention, sampling, or CUDA kernels.</span></p></li></ul><p><span>The unit test verifies that slot allocation and reuse order remain unchanged. Therefore, model output accuracy is not expected to change.</span></p><p><span>Local checks completed:</span></p><pre><code><span>PASS: Python syntax compilation for the modified source and test files
PASS: scripts/ci/check_registered_tests.py
PASS: git diff --check</span></code></pre><p><span>The PyTorch unit test is added to the existing registered CPU test suite and will be executed by GitHub CI:</span></p><pre><code><span>python3 -m unittest \
  test.registered.unit.mem_cache.test_mem_pool_host.TestLazyHostPoolRelease.test_logical_pool_lazy_release</span></code></pre><h2><span>Speed Tests and Profiling</span></h2><p><span>Before this change, every </span><code><span>LogicalHostPool.free(K)</span></code><span> may concatenate the released slots with an existing free list containing </span><code><span>F</span></code><span> entries:</span></p><pre><code><span>Free-list maintenance cost: O(F + K)</span></code></pre><p><span>After this change, </span><code><span>free()</span></code><span> only records the released tensor and updates a counter:</span></p><pre><code><span>Free-list maintenance cost: O(1)</span></code></pre><p><span>The larger concatenation is deferred until the primary free list cannot satisfy an allocation. Multiple releases can therefore be combined into one allocation-side merge instead of repeatedly copying the full free list.</span></p>
Behavior | Before | After
-- | -- | --
torch.cat during every free() | Yes | No
Primary allocation fast path | Tensor slice | Tensor slice
Released-slot accounting | Eager full-list rebuild | Append chunk and update counter
Large merge | Once per release | Only when released slots are needed
Page-alignment semantics | Preserved | Preserved
Allocation/reuse order | Primary free list, then released slots | Unchanged

<p><span>In the internal HiCache workload that exposed this issue:</span></p><ul><li><p><span>the pre-change eviction-and-retry path reached approximately </span><code><span>392.6–586.7 ms</span></code><span> average CPU latency;</span></p></li><li><p><span>the maximum observed latency was approximately </span><code><span>592 ms</span></code><span>;</span></p></li><li><p><span>after applying lazy release, repeated eager concatenations were removed from the </span><code><span>free()</span></code><span> hot path;</span></p></li><li><p><span>the amount of traffic required for HiCache warm-up was substantially reduced, although other cache-population costs can still require a smaller warm-up period.</span></p></li></ul><p><span>This matches the optimization strategy and performance rationale already validated for other host pools in #30658.</span></p><h2><span>Checklist</span></h2><ul><li><span><input></span><div><p><span>Format the code with </span><a href="https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit"><span>pre-commit</span></a><span>. Pending local dependency setup or GitHub CI.</span></p></div></li><li><span><input></span><div><p><span>Add unit tests according to the </span><a href="https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests"><span>unit-test guidance</span></a><span>.</span></p></div></li><li><span><input></span><div><p><span>Documentation update is not required because this change does not introduce or modify a user-facing interface.</span></p></div></li><li><span><input></span><div><p><span>Provide accuracy and performance analysis.</span></p></div></li><li><span><input></span><div><p><span>Follow the SGLang code-style guidance.</span></p></div></li></ul><h2><span>Review and Merge Process</span></h2><ol start="1"><li><p><span>Ping Merge Oncalls to start the process. See the </span><a href="https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process"><span>PR Merge Process</span></a><span>.</span></p></li><li><p><span>Get approvals from </span><a href="https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS"><span>CODEOWNERS</span></a><span> and other reviewers.</span></p></li><li><p><span>Trigger CI tests with the documented CI comments or contact an authorized user.</span></p></li><li><p><span>After CI is green and required approvals are collected, ask Merge Oncalls or a maintainer with write permission to merge the PR.</span></p></li></ol></body></html>## Motivation

This PR is a follow-up to [[#30658](https://github.com/sgl-project/sglang/pull/30658)](https://github.com/sgl-project/sglang/pull/30658).

PR #30658 defers host-pool free-list concatenation for `HostKVCache`-based pools. However, `LogicalHostPool` is a standalone allocator and does not inherit from `HostKVCache`, so it is not covered by that optimization.

`LogicalHostPool.free()` still eagerly rebuilds the complete free-slot tensor on every release:

```python
self.free_slots = torch.cat([self.free_slots, released_indices])
```

When HiCache uses a large logical host pool, `free_slots` can contain hundreds of thousands or millions of entries. Releasing a small number of slots therefore copies the entire existing free list. Repeated releases during cache eviction can introduce significant CPU stalls during HiCache cold fill and warm-up.

In an internal DSV4 HiCache workload, the allocator eviction-and-retry path showed representative average latencies of approximately `392.6–586.7 ms`, with a maximum close to `592 ms`. Profiling identified repeated large free-list concatenations as the dominant CPU-side overhead.

The issue was especially visible with a large logical cache capacity, but the underlying allocator problem is independent of any particular cache-sharding configuration.

## Modifications

Apply the lazy free-slot release strategy from #30658 to `LogicalHostPool`:

- Keep released slot tensors in `release_slots` instead of concatenating them immediately.
- Track pending released slots with `num_release_slots`.
- Include pending released slots in `available_size()`.
- Continue allocating from the primary `free_slots` tensor while it has sufficient capacity.
- Merge pending released chunks only when an allocation cannot be satisfied by the primary free list.
- Avoid recording empty releases.
- Reset all pending-release state in `clear()`.
- Preserve the existing page-alignment validation and slot reuse order.

The patch does not add runtime logging, configuration flags, or model-specific behavior.

Unit tests cover:

- deferred release without rebuilding the primary free list;
- allocation from the primary free list before merging pending slots;
- reuse of released slots after a deferred merge;
- merging multiple released chunks;
- empty releases;
- state reset through `clear()`;
- preservation of `LogicalHostPool` page-alignment checks.

## Accuracy Tests

This change only modifies CPU-side free-slot allocator metadata.

It does not modify:

- model weights or model forward execution;
- KV cache tensor contents;
- cache keys or page-table mappings;
- host-to-device or device-to-host data transfer;
- attention, sampling, or CUDA kernels.

The unit test verifies that slot allocation and reuse order remain unchanged. Therefore, model output accuracy is not expected to change.

Local checks completed:

```text
PASS: Python syntax compilation for the modified source and test files
PASS: scripts/ci/check_registered_tests.py
PASS: git diff --check
```

The PyTorch unit test is added to the existing registered CPU test suite and will be executed by GitHub CI:

```bash
python3 -m unittest \
  test.registered.unit.mem_cache.test_mem_pool_host.TestLazyHostPoolRelease.test_logical_pool_lazy_release
```

## Speed Tests and Profiling

Before this change, every `LogicalHostPool.free(K)` may concatenate the released slots with an existing free list containing `F` entries:

```text
Free-list maintenance cost: O(F + K)
```

After this change, `free()` only records the released tensor and updates a counter:

```text
Free-list maintenance cost: O(1)
```

The larger concatenation is deferred until the primary free list cannot satisfy an allocation. Multiple releases can therefore be combined into one allocation-side merge instead of repeatedly copying the full free list.

In the internal HiCache workload that exposed this issue:

- the pre-change eviction-and-retry path reached approximately `392.6–586.7 ms` average CPU latency;
- the maximum observed latency was approximately `592 ms`;
- after applying lazy release, repeated eager concatenations were removed from the `free()` hot path;
- the amount of traffic required for HiCache warm-up was substantially reduced, although other cache-population costs can still require a smaller warm-up period.

This matches the optimization strategy and performance rationale already validated for other host pools in #30658.

## Checklist

- [ ] Format the code with [[pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit)](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit). Pending local dependency setup or GitHub CI.
- [x] Add unit tests according to the [[unit-test guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests)](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Documentation update is not required because this change does not introduce or modify a user-facing interface.
- [x] Provide accuracy and performance analysis.
- [x] Follow the SGLang code-style guidance.

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [[PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [[CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS)](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with the documented CI comments or contact an authorized user.
4. After CI is green and required approvals are collected, ask Merge Oncalls or a maintainer with write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #31766142378](https://github.com/sgl-project/sglang/actions/runs/31766142378)
Latest PR Test (Extra): :white_check_mark: [Run #31766142172](https://github.com/sgl-project/sglang/actions/runs/31766142172)

### CODE REVIEW

**Comment by wangwenmingaa** [comment]
@hzh0425 could you please take a look at this PR when you have a chance?

This is a follow-up to #30658 and extends the deferred free-list concatenation optimization to LogicalHostPool, which is not covered by HostKVCache. In our DSV4 HiCache workload, the current eager concatenation caused allocator eviction-and-retry latency of roughly 393–587 ms, with peaks near 592 ms.

The change is focused and addresses a significant CPU-side bottleneck for large logical host pools. If the implementation looks good to you, I’d really appreciate your help reviewing and merging it. Thank you!

**Review by hzh0425 — APPROVED** [review]

**Comment by huangtingwei9988** [comment]
Thanks！
